File: 2023/day-1/zig/part2.zig
const std = @import("std");
const fs = std.fs;

pub fn main() !void {
    const file_name = "../input.txt";
    const buffer = try readFile(file_name, std.heap.page_allocator);

    const character_numbers = [_]*const []const u8{
        &"zero",
        &"one",
        &"two",
        &"three",
        &"four",
        &"five",
        &"six",
        &"seven",
        &"eight",
        &"nine",
    };

    var first_char: ?u8 = null;
    var second_char: ?u8 = null;
    var sum: u64 = 0;

    for (buffer, 0..) |char, index| {
        if (char == '\n') {
            if (second_char != null) {
                sum += asciiDigitToDigit(first_char.?) * 10 + asciiDigitToDigit(second_char.?);
            } else {
                sum += asciiDigitToDigit(first_char.?) * 10 + asciiDigitToDigit(first_char.?);
            }

            first_char = null;
            second_char = null;
            continue;
        }

        if (isNumberAtIndex(buffer, index, &character_numbers)) |val| {
            if (first_char == null) {
                first_char = val;
            } else {
                second_char = val;
            }
        }
    }

    std.debug.print("Sum: {}\n", .{sum});
}

fn isNumberAtIndex(buffer: []u8, buffer_index: usize, character_numbers: []const *const []const u8) ?u8 {
    const char_at_pos = buffer[buffer_index];

    if (char_at_pos == '\n') {
        return null;
    }

    if (isDigit(char_at_pos)) {
        return char_at_pos;
    }

    outer: for (character_numbers, 0..) |character_number, index1| {
        for (character_number.*, 0..) |char, index2| {
            if (buffer.len <= buffer_index + index2) {
                continue :outer;
            }

            if (buffer[buffer_index + index2] == '\n') {
                continue :outer;
            }

            if (buffer[buffer_index + index2] != char) {
                continue :outer;
            }
        }

        return @intCast(index1 + 48);
    }

    return null;
}

fn isDigit(char: u8) bool {
    if (char >= 48 and char <= 57) {
        return true;
    }

    return false;
}

fn asciiDigitToDigit(char: u8) u64 {
    return char - 48;
}

fn readFile(file_name: []const u8, allocator: std.mem.Allocator) ![]u8 {
    const file = try fs.cwd().openFile(file_name, .{});
    defer file.close();

    const file_stats = try file.stat();
    const file_size = file_stats.size;
    const buffer = try allocator.alloc(u8, file_size);
    _ = try file.readAll(buffer);
    return buffer;
}

File: 2023/day-01/part2.py

with open('input.txt', 'r') as file:
    input_text = file.read()

split_input = input_text.split('\n')
ready_list = list(filter(lambda x: x != '', split_input))

class CalibrationNumber:
    def __init__(self, index, number) -> None:
        self.index = index
        self.number = number
    
    def __str__(self) -> str:
        return f'({self.index}, {self.number})'

def get_numbers(string):
    string = string.lower()
    string_length = len(string)

    calibration_numbers = []

    for i in range(len(string)):
        if string[i].isdigit():
            calibration_numbers.append(CalibrationNumber(i, int(string[i])))
            continue
        
        if string_length - i + 1 >= 3 and string[i:i + 3] == 'one':
            calibration_numbers.append(CalibrationNumber(i, 1))
            continue
        
        if string_length - i + 1 >= 3 and string[i:i + 3] == 'two':
            calibration_numbers.append(CalibrationNumber(i, 2))
            continue
        
        if string_length - i + 1 >= 5 and string[i:i + 5] == 'three':
            calibration_numbers.append(CalibrationNumber(i, 3))
            continue
        
        if string_length - i + 1 >= 4 and string[i:i + 4] == 'four':
            calibration_numbers.append(CalibrationNumber(i, 4))
            continue
        
        if string_length - i + 1 >= 4 and string[i:i + 4] == 'five':
            calibration_numbers.append(CalibrationNumber(i, 5))
            continue
        
        if string_length - i + 1 >= 3 and string[i:i + 3] == 'six':
            calibration_numbers.append(CalibrationNumber(i, 6))
            continue

        if string_length - i + 1 >= 5 and string[i:i + 5] == 'seven':
            calibration_numbers.append(CalibrationNumber(i, 7))
            continue
        
        if string_length - i + 1 >= 5 and string[i:i + 5] == 'eight':
            calibration_numbers.append(CalibrationNumber(i, 8))
            continue
        
        if string_length - i + 1 >= 4 and string[i:i + 4] == 'nine':
            calibration_numbers.append(CalibrationNumber(i, 9))
            continue
    
    return calibration_numbers
    


def get_calibration_number(string):
    calibration_numbers = get_numbers(string)
    calibration_numbers.sort(key=lambda x: x.index)

    res = int(f'{calibration_numbers[0].number}{calibration_numbers[-1].number}')

    print(string)
    print(res)

    return res


res = sum(map(get_calibration_number, ready_list))
print(res)
